### PR TITLE
support for multiple profile failover urls

### PR DESCRIPTION
**Pull request by ajf8**
Hi,

We would like to make use of multiple failover URLs, by making the profile_failover setting comma seperated.

### CODE REVIEW

**Comment by ajf8** [comment]
test this please

**Comment by ned21** [comment]
OK to test

**Comment by stdweird** [comment]
@ajf8 can you update the config help with the new support https://github.com/quattor/CCM/blob/master/src/main/perl/CCfg.pm#L65
(and i still need to add you to all jenkins quattor repos, apologies for that)

**Comment by stdweird** [comment]
@ajf8 @ned21 another possiblility is make `profile` a comma separated list (and deprecate/get rid of the failover altogether)?
i also need to check if the config parser understands the `=s@` that converts commaseparated items in list automatically (see `ARGCOUNT_LIST` in http://search.cpan.org/~abw/AppConfig-1.56/lib/AppConfig.pm)

**Comment by stdweird** [comment]
@ajf8 to be able trigger PR builds, all you need to do is to join the quattor `developers` team (somehow).

**Comment by ajf8** [comment]
As suggested I've added a mention that profile_failover can be comma seperated.

Regarding making profile comma seperated instead, this was something we discussed here, and it was the initial plan. But the PROFILE_URL is used in a number of places in Fetch.pm which would probably need changing.

It seems to do some auto-detection of the format using this URL, so it would need to be changed to either check the first/last/all. The profile is downloaded to a filename which is the encoded profile URL. Adding the BASE_URL (if defined) to the profile URL would need to handle multiple values, and a bunch of error/debug messages which might need changing too.

This is all quite possible of course, but as a change it's a lot more involved than making profile_failover support multiple values, which is why I went for this approach.

File: src/main/perl/CCfg.pm [hunk 1]
```diff
@@ -62,7 +62,7 @@ Readonly::Array our @CONFIG_OPTIONS => (
     {
         option => 'profile_failover',
         suffix => '=s',
-        HELP => 'URL of profile to fetch when --profile is not available',
+        HELP => 'URL of profile to fetch when --profile is not available, can be comma seperated',
```
Comment (stdweird) [inline src/main/perl/CCfg.pm]: `seperated` -> `separated` ?

**Comment by stdweird** [comment]
@ajf8 ok, sounds good to me. if anyone want to fix this, they should open a new issue or PR to deprecate the failover setting.

the next issue however will be supporting this in `ncm-ccm`, in particular in the schema https://github.com/quattor/configuration-modules-core/blob/master/ncm-ccm/src/main/pan/components/ccm/schema.pan#L15 . making it a `type_hostURI[]` is not backwards compatible (and will need minor modification in the component). any toughts?

**Comment by ajf8** [comment]
Good point about the component. I've made a new CSV variant of the URI type in this branch here, so it should be backwards compatible. If you think this approach is OK I'll raise a PR. Thanks!

https://github.com/ajf8/configuration-modules-core/commit/9d3fad50e1a304ee676bf0ea3e576238c95da6d2

**Comment by stdweird** [comment]
@ajf8 i'd rather not have it done like that. the schema supports native lists, so , if possible we should use them.
my favourites are:
a. switch to `type_hostURI[]`, and have the component create the correct joined string. this is backwards imcompatible, so lots of other people will mind :smile: 

b. introduce new attribute `profiles_failover`, make it `type_hostURI[]` and add support for it in `ncm-ccm` and `Fetch`; and deprecate the `profile_failover`. this is probably the cleanest way forward (and correct way to do this kind of operation imho), but requires some more work.

c. use your approach, but without introducing a new generic type/function (it's the wrong location for such generic types and i'd rather not have more people use it :grin: ). e.g.
```pan
'profile_failover' ? string with {
    foreach(i; uri; split(",", SELF)) {
        if (!is_hostURI(uri)) { return(false); };
    };
    true;
}
```

i'll try to get some more opinions from the mailing list on this issue

**Comment by jrha** [comment]
I vote for option **a**, selfishly because we are not currently using it, so a backwards incompatible change will not hurt us.

**Comment by jrha** [comment]
@ajf8 any objection to option a?

**Comment by ajf8** [comment]
Hi James,

Sorry on the delay getting back. Yes I'm happy with that, I'll try to update this PR tomorrow and raise a PR for ncm-ccm.

I guess it's worth making the component .pm support both a string and a list. For us this is would be worth doing as the schema and perl will be updated at different times.

Thanks!

**Comment by jrha** [comment]
Sounds good, thanks!

**Comment by ajf8** [comment]
Rebased to fix conflicts with new unit tests in master, otherwise unchanged. Will send the components PR shortly.

**Comment by ajf8** [comment]
https://github.com/quattor/configuration-modules-core/pull/554

**Comment by stdweird** [comment]
with https://github.com/quattor/configuration-modules-core/pull/554 merged, this is good to go

**Pull request merged**